include/circt/Dialect/Moore/MooreOps.td

```diff
@@ -345,23 +345,26 @@ def NamedConstAttr : I32EnumAttr<"NamedConst", "elaboration-time constants",
 }
 
 def NamedConstantOp : MooreOp<"named_constant", [
-  SameOperandsAndResultType,
-  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>
+  FirstAttrDerivedResultType,
+  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>,
+  HasParent<"SVModuleOp">
 ]>{
   let summary = "An elaboration time constant expression";
   let description = [{
-    Constants are named data objects that never change. SystemVerilog provides 
+    Constants are named data objects that never change. SystemVerilog provides
     three elaboration-time constants: parameter, localparam, and specparam.
 
     See IEEE 1800-2017 § 6.20 "Constants".
   }];
-  let arguments = (ins StrAttr:$name, 
-                       NamedConstAttr:$kind, 
-                       UnpackedType:$value);
-  let results = (outs UnpackedType:$result);
+  let arguments = (ins StrAttr:$name,
+                       NamedConstAttr:$kind,
+                       AnyAttr:$value);
+  let results = (outs IntType:$result);
```

Comment:
Since you know that the `named_constant` will always return an integer, `$value` could be an `IntegerAttr` as well :smiley: